cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,387 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	// reference point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
 	}
-	if haveSave {
-		// 3.2a we have ubuntu-save, verify it as well
-		matches, err = disk.MountPointIsFromDisk(boot.InitramfsUbuntuSaveDir, diskOpts)
+
+	// 3. run the state machine logic for mounting partitions, this involves
+	//    trying to unlock then mount ubuntu-data, and then unlocking and
+	//    mounting ubuntu-save
+	//    see the state* functions for details of what each step does and
+	//    possible transition points
+
+	// TODO: should we just put a diagram here explaining the states too?
+
+	ctx := &recoverContext{
+		disk:          disk,
+		degradedState: recoverDegradedState{},
+	}
+
+	// first state to execute is to unlock ubuntu-data with the run key
+	machine := newStateMachine()
+	for {
+		final, err := machine.execute(ctx)
+		// TODO: consider whether certain errors are fatal or not
 		if err != nil {
 			return err
 		}
-		if !matches {
-			return fmt.Errorf("cannot validate boot: ubuntu-save mountpoint is expected to be from disk %s but is not", disk.Dev())
+		if final {
+			break
 		}
 	}
 
-	// 4. final step: copy the auth data and network config from
-	//    the real ubuntu-data dir to the ephemeral ubuntu-data
-	//    dir, write the modeenv to the tmpfs data, and disable
-	//    cloud-init in recover mode
-	if err := copyUbuntuDataAuth(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+	// 3.1 write out degraded.json
+	b, err := json.Marshal(ctx.degradedState)
+	if err != nil {
 		return err
 	}
-	if err := copyNetworkConfig(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+
+	// needed?
+	err = os.MkdirAll(boot.InitramfsHostUbuntuDataDir, 0755)
+	if err != nil {
 		return err
 	}
-	if err := copyUbuntuDataMisc(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+
+	err = ioutil.WriteFile(filepath.Join(boot.InitramfsHostUbuntuDataDir, "degraded.json"), b, 0644)
+	if err != nil {
 		return err
 	}
 
+	// 4. final step: copy the auth data and network config from
+	//    the real ubuntu-data dir to the ephemeral ubuntu-data
+	//    dir, write the modeenv to the tmpfs data, and disable
+	//    cloud-init in recover mode
+
+	// if we have the host location, then we were able to successfully mount
+	// ubuntu-data, and as such we can proceed with copying files from there
+	// onto the tmpfs
+	if ctx.degradedState.HostLocation != "" {
+		if err := copyUbuntuDataAuth(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+			return err
+		}
+		if err := copyNetworkConfig(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+			return err
+		}
+		if err := copyUbuntuDataMisc(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+			return err
+		}
+	} else {
+		// we don't have ubuntu-data host mountpoint, so we should setup safe
```

Comment:
in this mode, we still would run recovery snaps from the seed, so a device manufacturer could put in a recover-specific snap service which runs and does something, presumably the things it would do are:
- refresh snapd snap in recovery
- refresh the device recovery specific snap
- do something new that we haven't designed yet to recover the device for real